### PR TITLE
[Feature] support cloud native persistent index - add sstable

### PR DESCRIPTION
**Pull request by luohaha**
## Why I'm doing:
We are working on a new persistent index for cloud native primary key table. This new persistent index put it's data file on S3/hdfs, and have better performance and more adaptable to cloud-native environments.

## What I'm doing:
1. Build data format based on leveldb's sstable data format. So I copy some files from leveldb project (https://github.com/google/leveldb), and do some modifications based on them.
2. Add class `PersistentIndexSstable` in `storage/sstable/persistent_index_sstable.h/cpp`, it provides the ability to create sstable files, and to query (including multi-get and iterate by order).
3. Add function `MultiGet` in `storage/sstable/table.h/cpp`, It can support get multi keys at the same time.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/test/storage/sstable/persistent_index_sstable_test.cpp [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+                auto cur_val = UNALIGNED_LOAD64(iter->value().get_data());
+                if (cur_val > val) {
+                    it->second = iter->value().to_string();
+                }
+            }
+            iter->Next();
+        }
+    }
+    {
+        phmap::btree_map<std::string, std::string> map2;
```
Comment (TszKitLo40) [inline be/test/storage/sstable/persistent_index_sstable_test.cpp:160]: Why do we need map2?

File: be/src/storage/sstable/persistent_index_sstable.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    auto end_ts = butil::gettimeofday_us();
+    TRACE_COUNTER_INCREMENT("multi_get", end_ts - start_ts);
+    const auto& key_index_infos = key_indexes_info.key_index_infos;
+    for (size_t i = 0; i < key_index_infos.size(); ++i) {
+        if (!index_value_infos[key_index_infos[i]].empty()) {
+            IndexValueWithVerPB index_value_with_ver_pb;
+            if (!index_value_with_ver_pb.ParseFromString(index_value_infos[key_index_infos[i]])) {
+                return Status::InternalError("parse index value info failed");
+            }
+            if (version < 0 && index_value_with_ver_pb.values_size() > 0) {
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/sstable/persistent_index_sstable.cpp:71]: Keep it, will use it later

[the remark this answers is not in the document]
Comment (TszKitLo40) [reply be/src/storage/sstable/persistent_index_sstable.cpp:71]: I have removed the related logic from `LakePerisistentIndex`. I think we can add multi-version together later.

File: be/test/storage/sstable/persistent_index_sstable_test.cpp (hunk 1 above)
Comment (luohaha) [reply be/test/storage/sstable/persistent_index_sstable_test.cpp:160]: Because we will use map later for more check

Comment (TszKitLo40) [reply be/test/storage/sstable/persistent_index_sstable_test.cpp:160]: But map2 is not used now.

Comment (luohaha) [reply be/test/storage/sstable/persistent_index_sstable_test.cpp]: Done

File: be/test/storage/sstable/persistent_index_sstable_test.cpp [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+    ASSERT_OK(sstable::PersistentIndexSstable::build_sstable(map, file.get(), &filesz));
+    // 2. open sstable
+    std::unique_ptr<sstable::PersistentIndexSstable> sst = std::make_unique<sstable::PersistentIndexSstable>();
+    ASSIGN_OR_ABORT(auto read_file, fs::new_random_access_file(lake::join_path(kTestDir, filename)));
+    std::unique_ptr<Cache> cache_ptr;
+    cache_ptr.reset(new_lru_cache(100));
+    ASSERT_OK(sst->init(read_file.get(), filesz, cache_ptr.get()));
+
+    {
+        // 3. multi get with version (all keys included)
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/test/storage/sstable/persistent_index_sstable_test.cpp:235]: Done

**Review by decster — APPROVED** [review]

File: be/src/storage/CMakeLists.txt [hunk 4]
```diff
@@ -210,6 +210,20 @@ add_library(Storage STATIC
     lake/vertical_compaction_task.cpp
     lake/metacache.cpp
     lake/lake_primary_key_recover.cpp
+    sstable/block_builder.cpp
```
Comment (kevincai) [inline be/src/storage/CMakeLists.txt:213]: wondering if it worths a separate .a archive.

**Review by kevincai — APPROVED** [review]

File: be/src/storage/CMakeLists.txt (hunk 4 above)
Comment (luohaha) [reply be/src/storage/CMakeLists.txt:213]: I only want to use sstable format and do some hack on it. I don't think we can use it as a lib.

**Review by wyb — APPROVED** [review]

**Pull request merged**